src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -1198,6 +1199,12 @@ private void AbortStreams(int lastValidStream, Exception abortException)
 
                 CheckForShutdown();
             }
+
+            if (shouldInvalidate)
+            {
+                // Invalidate outside of lock to avoid race with HttpPool Dispose()
```

Comment:
It's more about maintaining a lock hierarchy, right? i.e. code holding the pool's lock might call into the connection and try to acquire the connection's lock, which means we mustn't hold the connection's lock while trying to acquire the pool's lock.

Reply:
yes, this is about two treads getting two locks in opposite order. I'll try to get with better explanation.